Beam/Include/Beam/Threading/Sync.hpp

```diff
@@ -27,6 +61,50 @@ namespace Threading {
       //! The type of mutex used.
       using Mutex = MutexType;
 
+      //! The type of the lock used for immutable access.
+      using ReadLock = Details::GetReadLock<Mutex>;
+
+      //! The type of the lock used for mutable access.
+      using WriteLock = Details::GetWriteLock<Mutex>;
+
+      //! The proxy for any type of lock in use.
+      class LockProxy {
+        public:
+          //! Construct an empty proxy.
+          LockProxy() = default;
+
+          //! Constructs a proxy to a lock.
+          /*!
+            \tparam The type of the lock.
+            \param lock The pointer to the lock.
+          */
+          template<typename T>
+          explicit LockProxy(T* lock)
+            : m_lock(lock) {}
+      
+          //! Locks the lock referenced.
+          void lock() {
+            std::visit(
+              [] (auto& ptr) {
+                ptr->lock();
+              });
+          }
+      
+          //! Unlocks the lock referenced.
+          void unlock() {
+            std::visit(
+              [] (auto& ptr) {
+                ptr->unlock();
+              });
+          }
+      
+        private:
+          using Variant = std::conditional_t<std::is_same_v<ReadLock,
+            WriteLock>, std::variant<WriteLock*>, std::variant<ReadLock*,
```

Comment:
Do not use a variant when ```ReadLock == WriteLock```. Instead use ```if constexpr``` within the ```lock```/```unlock``` methods.

Reply:
Just wondering, is there runtime overhead for using single-type variants?

Reply:
After testing it in www.godbolt.com , there is no over-head for single-type variants. Disregard the previous comment about using ``if constexpr```.

Reply:
To clarify, there is no over-head for single-type variants whose copy/move/assignment operations are ```noexcept```.